src/sources/journald.rs

```diff
@@ -295,16 +316,22 @@ where
                 }
 
                 saw_record = true;
-                if !self.units.is_empty() {
-                    // Make sure the systemd unit is exactly one of the specified units
-                    if let Some(unit) = record.get(&SYSTEMD_UNIT) {
-                        if !self.units.contains(unit) {
+
+                if let Some(unit) = record.get(&SYSTEMD_UNIT) {
+                    if !self.include_units.is_empty() {
+                        // Make sure the systemd unit is exactly one of the specified units
+                        if !self.include_units.contains(unit) {
                             continue;
                         }
-                    } else {
+                    }
+                    // Make sure the systemd unit is not specifically excluded
+                    if !self.exclude_units.is_empty() && self.exclude_units.contains(unit) {
                         continue;
                     }
+                } else if !self.include_units.is_empty() {
+                    continue;
                 }
```

Comment:
This looks good to me but is approaching the point where we should probably extract a function for more thorough testing.

I think there are also a few branch combinations here with redundant `is_empty() && contains()` checks. Seems like a prime candidate for a [decision table](https://www.hillelwayne.com/post/decision-tables/) :smile:

Reply:
Extracting the function makes sense. I'll try to apply the decision table methodology here, but I'm not sure how far I can go given that we want to avoid calling `contains` if at all possible (it has no internal optimizations for the empty table case).